### ISSUE TITLE
[Contributor Program][Task 18] Add a format-aware reward router with startup validation

### ISSUE THREAD

**Issue by TabsPhasers** [redai-infra/Relax#123]
## Claim and parallel-work notice

I claimed Task 18 of the Relax Contributor Program 2026 Cohort 1 as GitHub user `TabsPhasers` in [the contributor-program claim thread](https://github.com/redai-infra/Relax/issues/86#issuecomment-5078051789).

I am aware that another participant opened #97 and PR #118 for the same task. This is an independently developed proposal. Maintainers may redirect or consolidate the work.

## Goal and acceptance criteria

Implement a format-aware built-in reward router that:

- routes every sample in a mixed math + multiple-choice batch to the correct reward through explicit metadata or a conservative label matcher;
- supports a configurable fallback order while preserving explicit `--rm-type` as a compatible routing source;
- lets a built-in reward register its handler, execution mode, and optional matcher in one registry entry, without adding a reward-name branch to the generic router;
- warns when an invalid or conflicting higher-priority source falls back to a later valid source;
- fails visibly when no configured source can resolve a reward instead of silently treating a configuration error as a legitimate zero score;
- covers routing, fallback, conflicts, registry extension, mixed batches, and startup validation with CPU unit tests.

`Sample.label` is a ground-truth answer, not a reward-type field. Label routing therefore means conservative format recognition over existing labels; it does not introduce a new structured label schema or a natural-language classifier.

## Fixed repository and environment

| Item | Value |
| --- | --- |
| Repository | `https://github.com/redai-infra/Relax.git` |
| Baseline branch | `main` |
| Baseline commit | `cfda05351d4b137f60e2b71d048beab1316d9d9f` |
| Development branch | `contrib/task-18-format-aware-reward-router` |
| Operating system | macOS 15.7.4, arm64 |
| CPU / memory | Apple M2, 8 cores / 24 GB |
| Python | 3.12.9 |
| PyTorch / Ray | 2.13.0 / 2.56.0 |
| CUDA / GPU | unavailable / none; this task is CPU-testable |
| Environment | local virtual environment; no new dependency or container image |

This is a correctness task rather than a throughput benchmark, so there is no performance target or GPU measurement.

## Complete validation commands

```bash
# Run from the Relax repository root on the fixed development branch.
export CUDA_VISIBLE_DEVICES=""

python -m pytest -v \
  tests/engine/rewards/test_reward_router.py \
  tests/engine/rewards/test_reward_worker.py \
  tests/engine/rollout/test_data_source.py

pre-commit run --all-files --show-diff-on-failure
git diff --check
```

No training script or public CLI argument is added. Routing configuration reuses the existing `--custom-config-path` YAML mechanism.

## Proposed contract

`custom_rm_path` remains an unconditional override. The default built-in source order is:

```text
valid metadata["rm_type"] > exactly one label matcher > valid explicit --rm-type
```

The three built-in sources can be reordered in existing custom YAML, for example:

```yaml
reward_route:
  priority: [metadata, rm_type, label]
```

Every registered reward remains directly selectable by `metadata["rm_type"]` or `--rm-type`. Supplying an optional matcher additionally makes that reward eligible for label routing. Matcher implementations belong to their concrete reward registrations; the generic router only calls the registered predicate.

If an unknown or conflicting higher-priority source is recovered by a later valid source, routing continues with a warning. If every source fails, standard datasets abort during startup validation and unvalidated runtime samples preserve the existing exception behavior.

## Files and scope

Expected implementation surfaces:

- `relax/engine/rewards/reward_router.py`
- `relax/engine/rewards/__init__.py`
- `relax/engine/rollout/data_source.py`
- `relax/distributed/ray/placement_group.py`
- targeted reward-router, reward-worker, and data-source tests
- `docs/zh/guide/customize-training.md`

Non-goals:

- no reward formula or training-algorithm changes;
- no changes to worker counts, batching, scheduling, resource allocation, or dataset schemas;
- no new CLI argument or dependency;
- no requirement that third-party data sources implement the built-in startup preflight;
- no changes to `custom_rm_path`, `boxed_`, or lazy optional-scorer imports.

## Delivery

Delivery method: this Issue plus PR #125. The PR contains the implemented design, file-level review guide, compatibility notes, and validation evidence.

**Comment by TabsPhasers** [comment]
Development branch: [`TabsPhasers:contrib/task-18-format-aware-reward-router`](https://github.com/TabsPhasers/Relax/tree/contrib/task-18-format-aware-reward-router). It currently points to upstream `main` at `cfda05351d4b137f60e2b71d048beab1316d9d9f` and intentionally contains no implementation commits yet. GitHub denied the native `CreateLinkedBranch` operation for this external contributor account, so the branch association is recorded here instead.
